File: src/java/magma/compile/rule/divide/MutableDividingState.java
package magma.compile.rule.divide;

import jvm.collect.list.Lists;
import magma.collect.list.List_;
import magma.option.Option;
import magma.option.Tuple;

public class MutableDividingState implements DividingState {
    private final List_<Character> queue;
    private List_<String> segments;
    private StringBuilder buffer;
    private int depth;

    public MutableDividingState(List_<Character> queue) {
        this(queue, Lists.empty(), new StringBuilder(), 0);
    }

    public MutableDividingState(List_<Character> queue, List_<String> segments, StringBuilder buffer, int depth) {
        this.queue = queue;
        this.segments = segments;
        this.buffer = buffer;
        this.depth = depth;
    }

    @Override
    public DividingState append(char c) {
        buffer.append(c);
        return this;
    }

    @Override
    public boolean isLevel() {
        return depth == 0;
    }

    @Override
    public DividingState exit() {
        this.depth = depth - 1;
        return this;
    }

    @Override
    public DividingState enter() {
        this.depth = depth + 1;
        return this;
    }

    @Override
    public DividingState advance() {
        this.segments = segments.add(buffer.toString());
        this.buffer = new StringBuilder();
        return this;
    }

    @Override
    public List_<String> segments() {
        return segments;
    }

    @Override
    public boolean isShallow() {
        return depth == 1;
    }

    @Override
    public Option<Tuple<Character, DividingState>> append() {
        return pop().map(tuple -> {
            return new Tuple<>(tuple.left(), tuple.right().append(tuple.left()));
        });
    }

    @Override
    public Option<Tuple<Character, DividingState>> pop() {
        return queue.popFirst().map(tuple -> new Tuple<>(tuple.left(), new MutableDividingState(tuple.right(), segments, buffer, depth)));
    }

    @Override
    public Option<DividingState> appendAndDiscard() {
        return append().map(Tuple::right);
    }
}

File: src/clang/magma/compile/rule/divide/MutableDividingState.c
#include "MutableDividingState.h"
int __lambda0__(){return (tuple.left(), tuple.right().append(tuple.left()));
}
int __lambda1__(){return (tuple.left(), MutableDividingState(tuple.right(), segments, buffer, depth));
}
public MutableDividingState(List_<char> queue){this(queue, Lists.empty(), StringBuilder(), 0);
}
public MutableDividingState(List_<char> queue, List_<String> segments, StringBuilder buffer, int depth){this.queue = queue;
        this.segments = segments;
        this.buffer = buffer;
        this.depth = depth;
}
DividingState append(char c){buffer.append(c);return this;
}
int isLevel(){return depth == 0;
}
DividingState exit(){this.depth = depth - 1;return this;
}
DividingState enter(){this.depth = depth + 1;return this;
}
DividingState advance(){this.segments = segments.add(buffer.toString());
        this.buffer = new StringBuilder();return this;
}
List_<String> segments(){return segments;
}
int isShallow(){return depth == 1;
}
Option<Tuple<char, DividingState>> append(){return pop().map(__lambda0__);
}
Option<Tuple<char, DividingState>> pop(){return queue.popFirst().map(__lambda1__);
}
Option<char> peek(){return queue.findFirst();
}
Option<DividingState> appendAndDiscard(){return append().map(Tuple.right);
}
